### PR TITLE
Support boolean attention mask in Attention(23) CUDA - MHA case

### PR DESCRIPTION
**Pull request by titaiwangms**
Replace and reland https://github.com/microsoft/onnxruntime/pull/27129

Comparison between this PR approach and inline in softmax

## Tradeoffs

| Category | Pre-conversion (current) | Inline in softmax |
| :--- | :--- | :--- |
| **Memory** | Extra buffer ($num\_elements \times sizeof(T)$) | None — reads 1-byte bool directly |
| **Kernel launches** | +1 simple elementwise kernel | Zero extra |
| **Code complexity** | 3 files, ~40 lines added | 6+ kernel templates, macros, dispatch logic, data structs |
| **Risk** | Low — softmax path untested | High — modifying battle-tested softmax kernels used by MHA + GQA contrib ops |
| **Perf impact** | Negligible — mask is small vs. QKV; conversion is memory-bound and fast | Slightly better theoretical bandwidth |
| **Maintainability** | Clean separation of concerns | Adds template dimension across all softmax variants |

---

This pull request enhances the ONNX Runtime CUDA Attention operator to support boolean attention masks (bool masks) in the Multi-Head Attention (MHA) path, converting them to additive attention bias on the GPU. It also improves test coverage to ensure correctness and parity with the CPU implementation. The main changes include implementing a CUDA kernel for mask conversion, updating the operator logic to handle bool masks, clarifying broadcasting rules, and adding comprehensive unit tests.

**CUDA Attention Operator Improvements:**

* Implemented a CUDA kernel (`LaunchConvertBoolMaskToAttentionBias`) that converts boolean attention masks to additive bias (True → 0.0, False → mask_filter_value) for the MHA path, ensuring efficient GPU execution. [[1]](diffhunk://#diff-00f7d49ccee44f1573357c07633bd03f21b9c2e1b1617c7a6a878a79ee6a6e49R148-R187) [[2]](diffhunk://#diff-8aa9a15a92d7dc138346dce5de055911895d940ba2183b4ba45bd95ac0e5bfc9R55-R66)
* Updated `attention.cc` to use this kernel, correctly handle bool masks in the MHA path, and clarified the broadcasting logic and mask shape interpretation for both GQA and MHA. [[1]](diffhunk://#diff-0701e4cc6d4951894ae1a60f35c1e6c0f69ba7595f896a23c8f5ed7265eab4ffR6) [[2]](diffhunk://#diff-0701e4cc6d4951894ae1a60f35c1e6c0f69ba7595f896a23c8f5ed7265eab4ffR380-R383) [[3]](diffhunk://#diff-0701e4cc6d4951894ae1a60f35c1e6c0f69ba7595f896a23c8f5ed7265eab4ffL514-L522) [[4]](diffhunk://#diff-0701e4cc6d4951894ae1a60f35c1e6c0f69ba7595f896a23c8f5ed7265eab4ffL549-R557) [[5]](diffhunk://#diff-0701e4cc6d4951894ae1a60f35c1e6c0f69ba7595f896a23c8f5ed7265eab4ffR595-R616)

**Testing and Documentation Enhancements:**

* Added new test cases and a dedicated test class to validate the correctness of boolean mask handling in the MHA path, ensuring parity with the CPU implementation for 2D, 3D, and 4D mask shapes. [[1]](diffhunk://#diff-801fbbcf2537e8e13a0202e6a0f7e88c56ab5aa72d17d949a5556355694b2b2dR563-R725) [[2]](diffhunk://#diff-801fbbcf2537e8e13a0202e6a0f7e88c56ab5aa72d17d949a5556355694b2b2dR893-R922)
* Improved comments and documentation in both code and tests to clarify ONNX broadcasting rules and mask shape expectations for different attention paths. [[1]](diffhunk://#diff-4ed1461afda0d3804a61ba95a64b2a84d0c1395f9c887d1a3fdfed914ade22c1L208-R221) [[2]](diffhunk://#diff-801fbbcf2537e8e13a0202e6a0f7e88c56ab5aa72d17d949a5556355694b2b2dR35)

**Test Coverage and Reliability:**

* Enabled CUDA-based tests for boolean mask scenarios previously only tested on CPU, and adjusted test logic to ensure correct handling of edge cases (e.g., all-false masks). [[1]](diffhunk://#diff-3ff6dfa2ce407ae0073009174c37d1756509e8bbc434dee7c44cd55a996bb777L477-R480) [[2]](diffhunk://#diff-3ff6dfa2ce407ae0073009174c37d1756509e8bbc434dee7c44cd55a996bb777L620-R623)

These changes make the CUDA Attention operator more robust and feature-complete, aligning its behavior with the CPU implementation and ONNX specifications.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/llm/attention.cc [hunk 1]
```diff
@@ -582,8 +578,26 @@ Status Attention<T>::ComputeInternal(OpKernelContext* context) const {
...
+        // Use native CUDA types for the kernel (OrtToCudaType maps BFloat16 -> __nv_bfloat16)
+        using NativeCudaT = typename onnxruntime::cuda::OrtToCudaType<T>::type;
+        int64_t num_elements = attn_mask->Shape().Size();
+        converted_mask_buffer = GetScratchBuffer<void>(num_elements * sizeof(NativeCudaT), context->GetComputeStream());
+        auto cuda_stream = static_cast<cudaStream_t>(context->GetComputeStream()->GetHandle());
+        ORT_RETURN_IF_ERROR(LaunchConvertBoolMaskToAttentionBias<NativeCudaT>(
+            attn_mask->Data<bool>(),
+            reinterpret_cast<NativeCudaT*>(converted_mask_buffer.get()),
+            num_elements,
+            contribop_parameters.mask_filter_value,
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxruntime/core/providers/cuda/llm/attention.cc:609]: Fixed

**Comment by justinchuby** [comment]
## 🔍 Independent Review by a Team of Copilots

**Review team**: 2 Architects (claude-opus-4.6), 1 Code Reviewer (gemini-3-pro-preview), 1 Critical Reviewer (claude-sonnet-4.6) — coordinated by a Copilot Project Lead

### Verdict: ✅ Conditional Approve

Excellent PR — clean design, great documentation, well-scoped change. One blocking issue to fix, two non-blocking suggestions.

---

### 🔴 BLOCKING: Grid Size Overflow (H1)

**File**: `attention_mask_impl.cu` — `LaunchConvertBoolMaskToAttentionBias`

`blocks` is `int64_t` but cast to `unsigned int` for the CUDA `<<<grid, block>>>` launch. CUDA `gridDim.x` max is 2³¹−1. For a 4D mask on a long-context model (e.g., shape `[1, 32, 128K, 128K]`), blocks can exceed this limit, causing **silent data corruption** — only a subset of mask elements would be converted.

**Suggested fix** — grid-stride loop (common CUDA best practice):
```cuda
__global__ void ConvertBoolMaskToAttentionBiasKernel(...) {
  for (int64_t idx = static_cast<int64_t>(blockIdx.x) * blockDim.x + threadIdx.x;
       idx < num_elements;
       idx += static_cast<int64_t>(gridDim.x) * blockDim.x) {
    attention_bias[idx] = attn_mask[idx] ? T(0.0f) : T(mask_filter_value);
  }
}
```
And cap the grid: `min(blocks, 65535u)` or similar.

---

### 🟡 Non-blocking suggestions

**M1 — Unused include** (`attention.cc:6`):
`#include "core/providers/cpu/llm/attention.h"` — no symbols from this header are used in the changed code. Creates an unnecessary CPU→CUDA provider dependency. Consider removing it, or using `mask_filter_value<T>()` from it instead of reimplementing with `std::numeric_limits<T>::lowest()` (the helper may have type-specific specializations).

**M2 — Test coverage gaps** (`test_attention.py`):
Python tests only cover fp16 with `batch_size=2` and a single config. Consider adding:
- fp32 test method (kernel is instantiated for float/half/bfloat16 but only half is tested)
- `batch_size=1` config (exercises broadcasting edge case)
- All-true mask edge case

---

### ✅ Confirmed Safe

We investigated several potential concerns and confirmed they are **not issues**:

| Concern | Status | Evidence |
|---------|--------|----------|
| `mask_filter_value` change from `-10000` to `lowest()` causing regression | **Safe** | Traced full call chain: value is only consumed by `ComputeSoftmaxWithRawMask`, which requires `data.mask_index != nullptr`. The new Attention op sets `mask_index = nullptr`, so the value is never consumed by softmax in non-bool paths. |
| Numerical stability / NaN from overflow | **Fixed** | Uses `std::numeric_limits<T>::lowest()` → `-65504` for fp16 (finite, not `-inf`). Copilot review fix correctly applied. |
| Buffer lifetime / use-after-free | **Safe** | `converted_mask_buffer` lives at function scope; all GPU work on same CUDA stream with stream-ordered allocation. |
| CPU/CUDA parity | **Achieved** | CPU path already uses `std::numeric_limits<T>::lowest()` via `mask_filter_value<T>()`. |

---

### 👏 Highlights

- **Broadcasting documentation** (lines 516–527) is outstanding — the right-aligned notation with examples is best-in-class
- **Pre-conversion architecture** is the right call — low risk, clean separation, negligible perf overhead vs. modifying battle-tested softmax kernels
- **GQA vs MHA asymmetry** is well-documented and architecturally justified
- **Kernel quality** is excellent — clean, focused, properly templated

---

*Review performed by a team of Copilots — 4 specialist agents with different models and perspectives, coordinated by a Copilot Project Lead. Findings represent cross-validated consensus.*

**Comment by titaiwangms** [comment]
A 4D mask [1, 32, 128K, 128K] = 32 × 128K × 128K ≈ 537 billion elements. At 1 byte per bool, that's 537 GB — it wouldn't even fit in GPU memory. The max gridDim.x
  is 2^31 - 1 ≈ 2.1 billion, and with 256 threads per block that covers ~550 billion elements, which is coincidentally close to the limit.

  But practically, even a 4D mask [8, 32, 8192, 8192] (very generous) = ~17 billion elements = ~17 GB of bool, which is already well beyond what any real model
  would use for an attention mask. With 256 threads/block, blocks = 17B / 256 ≈ 66M, well under the 2.1B grid limit.

  That said, the fix is trivially correct and costs nothing at runtime, so it's a reasonable defensive improvement. It's up to you — both "accept and fix" (takes 2
  lines) and "push back as impractical" are valid.

**Comment by justinchuby** [comment]
> A 4D mask [1, 32, 128K, 128K] = 32 × 128K × 128K ≈ 537 billion elements. At 1 byte per bool, that's 537 GB — it wouldn't even fit in GPU memory. The max gridDim.x is 2^31 - 1 ≈ 2.1 billion, and with 256 threads per block that covers ~550 billion elements, which is coincidentally close to the limit.
> 
> But practically, even a 4D mask [8, 32, 8192, 8192] (very generous) = ~17 billion elements = ~17 GB of bool, which is already well beyond what any real model would use for an attention mask. With 256 threads/block, blocks = 17B / 256 ≈ 66M, well under the 2.1B grid limit.
> 
> That said, the fix is trivially correct and costs nothing at runtime, so it's a reasonable defensive improvement. It's up to you — both "accept and fix" (takes 2 lines) and "push back as impractical" are valid.

Should have realized that 😅  AI team failed.
